PR: Add current version of docs style guide (#106)
spyder-ide/spyder-docs: STYLEGUIDE.md

# Spyder Documentation Style Guide:

## General

> Must (not), should (not), recommended, suggested, may, and can generally follow IETF conventions, with some additional differentiation:

Author: https://ietf.org/

> Must (not), should (not), recommended, suggested, may, and can generally follow IETF conventions, with some additional differentiation:

Author: All internet standards (RFCs) follow this along with numerous others.

- **Must** (not) clauses are hard requirements that must always be observed
- **Should** (not) clauses can only be disregarded in a particular instance with strong justification, or if it is demonstratively highly impractical to follow for a specific case
- (Not) **Recommended** and **Preferred** clauses should be followed where practicable; a brief rationale should be provided for not doing so
- **Suggested** clauses indicate a convention, approach or guideline that may be helpful to follow, but has no binding effect or justification requirements
- **May** or **Can** clauses designate explicitly permitted guidelines, actions or conventions, often for clarity, that are allowed but have no weight of preference on them (and may even be gently discouraged with a suggested or preferred alternative)

> 

Author: Its just any clause that uses the term, it can be anything. The style guide currently doesn't actually use them thoroughly because I never did the polish pass I planned to long ago before submitting this. Since this section is only normative to the style guide itself rather than to the actual docs, and not consistently employed as of now, I could just remove it.

## Structure

- In general, follow the [OpenStack Heading Guidelines](https://docs.openstack.org/doc-contrib-guide/writing-style/general-writing-guidelines.html) with regard to sections and headings; in particular, section and subsection titles delineating overall categories of tasks/topics and be in gerund (i.e. with "ing", e.g. "Editing code", "Debugging files"), while those concerning a concept or reference topic should be a noun (e.g. "General considerations", "Requirements") and lowest level topic/task titles should generally be in imperative ("Create a code cell", "Get help", etc) where reasonable

> - Heading case: Page titles (L1 headings) should use Title Case, all others should use Sentence case

Author: @juanis2112 ^

> - 1 L1 Heading for page titles: Use ``######`` with overline; on first line of file; keep to a few words

Author: ```suggestion
- One L1 Heading, used for page titles: Use ``######`` with overline; on first line of file; keep to a few words
```

Its to make explicit that there should only be one per document, but I've rewritten it to be clearer.

- L2 Headings for section heads: Use ========= with overline; 3 blank lines before them, will appear in sidebar so keep them short

> - L3 headings for subheads: Use ~~~~~~, 2 blank lines,

Author: ```suggestion
- L3 headings for subheads: Use ~~~~~~, 2 blank lines
```

Good catch. There is a lot of sloppiness here since I never did the big polish pass I was going to.

> - L4 Headings for sub-subheads if truly necessary: Use ---------, 2 blank lines, keep to a minimum (consider a flatter structure or breaking into multiple files)

Author: No, flatten would be a grammar error here, while flatter is correct. The former is an adjective (meaning "more flat") while the latter is a verb.

- Related components at the bottom: L3 heading, bullets, 2 blank lines before, try for 2-4 links if included
- All files with Related components should contain at least one L2 section; if only one, can be  entitled Overview
- Typically, for files with multiple sections or longer than one page, a summary section of a few lines length is helpful before the first section, summarizing the functionality discussed in the section and its purpose
- Name of component/topic should be in first sentence, Title Case, bold on first use

## Technical

- Markdown, Github-flavored (``.md``) for any non-documentation rendered text files in the root of the repo, e.g. ``README``, ``CONTRIBUTING``, etc.
- reStructuredText, (reST, ``.rst``) for documentation format, i.e. any user-visible files in the ``/doc`` directory

> - Where not otherwise contradicted by this guide or specific individual requirements, a subset of Markdown corresponding to that cross-compatible with reST should be preferred where possible; e.g. ``*italic*`` vs. ``_italic_``, ``*`` for bulleted lists, double backtick for inline code vs. single, etc.

Author: I've addressed this before, at some length. For Markdown in the Spyder repo it may be justifiable since it is less likely to have to interoperate with the reST docs, Markdown is the common format, you have to edit it frequently and edits are much more short-form. However, given none of this is the case (and noting that on an US English keyboard layout, which would be expected to be used/switched to when writing the long-form English prose on this repo, the additional time spent is minuscule relative to the substantial overhead converting back and forth if needed, even with tools), I judged it to be worthwhile and the content of this repo is all formatted accordingly.

If you and @juanis2112 disagree and want to change it, that's fine, but the style guide is meant to reflect the current state of the docs and repo, where this is the case. Once this is committed, if you want to change it, it can then be changed repo-wide and in the style guide, and the changes submitted as an atomic PR, so its clear.

> - Documentation should build with no warnings with ``Sphinx`` in nitpicky mode

Author: I did add it to Travis, it builds the docs in nitpicky mode and fails if so. However, that's all buried in a config file, putting it here makes it normative, persistent, and serves as a reminder for users to ensure their changes build.

- Lower case filenames, rst extension, hyphens as space deliminators, short but descriptive
- End files with a newline (blank line)
- 1 blank line after all headings and before and after paragraphs, directives and "|"s

> - Use Line Feed (LF, ``\n``) as the End of Line character on all platforms

Author: Yes, of course. Again, like the above, this makes it normative, persistent and avoids any confusion over why line endings were normalized.

- Use UTF-8 for all files (with no BOM, as dictated by the UTF-8 standard); do not use Windows Notepad as this adds a BOM and can otherwise corrupt files easily

> - Indicate special characters (e.g. em dash, en dash, etc) by simply including them inline with Unicode

Author: Instead of using escape sequences, simply use the actual UTF-8 character

> - Indicate special characters (e.g. em dash, en dash, etc) by simply including them inline with Unicode

Author: ```suggestion
- Use standard UTF-8 characters instead of escape sequences for non-ASCII values (e.g. em dash, en dash, etc)
```

- Suggested to avoid non-ASCII characters where minimal downside/need exists, and recommended to keep used characters within the Latin-1 set where an additional range is not required by the specific usage, or an existing proper name or other identifier uses them
- Spaces, no tabs for indentation and otherwise

> - Natural indents (3 spaces) for directives, 4 spaces for further indents beyond that in code

Author: Because that's the "natural" visual indent level and matches most other sources and style guides. It can be changed later, but its what's implemented now repo-wide.

- No hard wrap, put each sentence on its own line, and try to keep them under 180 characters
    - Wrap code blocks at a width of 70 characters from the page margin for readability.
- Place section links above the section with one blank line between them and the section, and two blank lines between sections (as appropriate for that section level) between the link and the preceding text.
- Always use single backticks around link titles

> - Eliminate any trailing whitespace before commiting

Author: Not without [some nasty hacks](https://stackoverflow.com/questions/591923/make-git-automatically-remove-trailing-whitespace-before-committing). A much better solution is pre-commit hooks to enforce a lot of this stuff, but we'd have to get them set up and ensure everyone uses them. We could use [pre-commit](https://pre-commit.com/) which seems like a pretty nice and comprehensive framework for them, though someone would actually have to implement it.

> - Eliminate any trailing whitespace before commiting

Author: Right now I just have a bunch of small bash scripts that checks most of these things, but they are pretty horribly written and take some time to run on larger codebases.

- Try to keep individual files to under around ~500 lines; split into separate topics beyond that

## reST Style Elements

> - Filenames/paths: Use :file: and ``/`` for deliminators e.g. :file:`path/to/test.py`. Example filenames should be all lowercase with no spaces or special characters unless called for by the application; real filenames should follow their actual case. Use {descriptive-name} for placeholders.

Reviewer: Is `deliminators` a word, or you meant `delimiters` here?

Author: ```suggestion
- Filenames/paths: Use :file: and ``/`` for delimiters e.g. :file:`path/to/test.py`. Example filenames should be all lowercase with no spaces or special characters unless called for by the application; real filenames should follow their actual case. Use {descriptive-name} for placeholders.
```

"deliminator" actually is a word, but delimiter was what I meant. Good catch, thanks.

> - Keyboard Shortcuts: Use :kbd: to format, Ctrl/Alt/Shift as abbreviations (instead of Command/Option), ``-`` with no spaces to separate, and title case (including capital letters for letter keys; always include Shift explicitly if needed) e.g. :kbd:`Ctrl-Alt-Shift-P`. Use the key name or symbol as printed on a standard keyboard. For any configurable shortcut, you should include "by default" to indicate this, and must include at least one another means the command can be accessed if available (e.g. via a menu or toolbar), which should be done for all shortcuts and should include all means of executing the command, if multiple others exist.

Reviewer: another -> other ?

Author: ```suggestion
- Keyboard Shortcuts: Use :kbd: to format, Ctrl/Alt/Shift as abbreviations (instead of Command/Option), ``-`` with no spaces to separate, and title case (including capital letters for letter keys; always include Shift explicitly if needed) e.g. :kbd:`Ctrl-Alt-Shift-P`. Use the key name or symbol as printed on a standard keyboard. For any configurable shortcut, you should include "by default" to indicate this, and must include at least one other means the command can be accessed if available (e.g. via a menu or toolbar), which should be done for all shortcuts and should include all means of executing the command, if multiple others exist.
```

Yup, it actually took me a minute to figure out that one myself, heh.

> - GUI elements (names of things in the UI): Use ``:guilabel:`` , e.g. ``:guilabel:`Name```; make sure to use the exact name and capitalization as printed in the UI, except for the proper names of Spyder panes (i.e. those listed in the ``View > Panes`` menu), which should always be written in Title Case ("History Log", "Project Explorer", "Find in Files", etc. as they are proper names). Also, for consistency and clarity, they should be referred to as panes, rather than modules, plugins, panels, etc.; Find in Files may be called a "tool" if appropriate to the context.

Reviewer: * There are three backticks at the end of this expression:

      ``:guilabel:`Name```

* "History Log", "Project Explorer", "Find in Files": these examples don't apply anymore. Please change them to "Variable Explorer", "IPython Console".

Author: ```suggestion
- GUI elements (names of things in the UI): Use ``:guilabel:`` , e.g. ``:guilabel:`Name` ``; make sure to use the exact name and capitalization as printed in the UI, except for the proper names of Spyder panes (i.e. those listed in the ``View > Panes`` menu), which should always be written in Title Case ("Variable Explorer", "Code Analysis", "IPython Console", etc. as they are proper names). Also, for consistency and clarity, they should be referred to as panes, rather than modules, plugins, panels, etc.; Find in Files may be called a "tool" if appropriate to the context.
```

> There are three backticks at the end of this expression:

Yes, which is the correct number. The first is the closing backtick for `Name`, and the second two are the closing backticks for the inline verbatim text. Apparently Github's GFM parser doesn't render this correctly so I've added a space to fix that.

> Please change them to "Variable Explorer", "IPython Console".

Fixed to three up to date examples. Ironically, I just noticed that in the Spyder UI, two of those I mentioned are Sentence case, while one is Title Case, and the rest of the UI text isn't much better. This is really my fault for never doing that polish pass I was going to...I can get back to it some day.

> - Menu items and preference panes: Use :menuselection: to format and ``-->`` to separate, e.g. `` :menuselection:`Project --> New Project` ``, make sure to use the exact name and capitalization of each level element

Reviewer: * Why not use `>` instead of `-->` here? That's the convention I've been following for years now.
* Also, I think you should use a dot instead of a comma here:

      :menuselection:`Project --> New Project` ``. Make sure ...

Author: Because `-->` is the correct delimiter for `:menuselection:` and `>` will not work, per the [Sphinx docs](https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html#role-menuselection). And also, again, it is what is implemented across all the docs; any changes should be made as a further PR that changes it both here and there.

Author: > Also, I think you should use a dot [sic] instead of a comma here:

Correct, you've detected a comma splice. Also FYI, in English its called a "period" in this context (in British English, it would be a "full stop").

```suggestion
- Menu items and preference panes: Use :menuselection: to format and ``-->`` to separate, e.g. `` :menuselection:`Project --> New Project` ``. Make sure to use the exact name and capitalization of each level element
```

- Short commands/lines of code (under 20-30 characters), use `` , e.g. ``code``

> - For any code longer than one line, ~20-30 characters, or that is otherwise important to emphasize rather than just mention, use ``code-block::`` tags, closed with ``end``, and specify the programming language (bash for command line, python for python, ini for configuration files, etc.). PEP 8 should be followed for Python code, with all "recommended"/"preferred"/"suggested" guidelines being treated as *should* i.e. only broken with appropriate justification in a situation requiring them; similarly, PEP 257 for docstrings where needed. Additionally, all code must be valid and idiomatic Python 3 unless specifically meant to illustrate a distinction with Python 2, and code should for the time being fall under the common subset supported by Spyder's supported platforms (currently, Python 2.7 and Python 3.4-3.6). Wrap code blocks at a width of 70 characters from the page margin for readability.

Reviewer: Python ~~3.4-3.6~~ 3.5-3.7

Author: Does Spyder still not support 3.8 yet?

- Internal links to sections should never rely on implicit references, since they may change with section titles or other re-organization. Instead, always use explicit references above the target with one blank line between; keep ref names short, use ``-`` to replace spaces, use unique and distinctive names and append ``-ref`` to the end of the name for clarity.

> - All external links should be inline (not bare) unless specifically needed to illustrate some property of the link itself, should be https if available, and should, generally, be written as a reference with `` `Test Link Name`_ `` and then included in full at after the end of the paragraph in which it appears, e.g. `` .. _Test Link Name: https://www.example.com/ `` . A block of links should be separated from paragraphs above and below by one blank line, with no blank lines between links. Make sure to check that your links, as the build will automatically check for broken ones. Strip anything unnecessary (parameters, referrers, etc) off the end. As the base docs are in US English, any links should be to the English, desktop version of the site.

Reviewer: * ... then included in full ~~at~~ after the end ...
* Make sure to check that your links <something text seems to be missing here?>, as the build will automatically check for broken ones.
* I don't understand what you mean by this: "Strip anything unnecessary (parameters, referrers, etc) off the end. As the base docs are in US English, any links should be to the English, desktop version of the site."

Author: ```suggestion
- All external links should be inline (not bare) unless specifically needed to illustrate some property of the link itself, should be HTTPS if available, and should, generally, be written as a reference with `` `Test Link Name`_ `` and then included in full after the end of the paragraph in which it appears, e.g. `` .. _Test Link Name: https://www.example.com/ `` . A block of links should be separated from paragraphs above and below by one blank line, with no blank lines between links. Make sure to check your links, as the build will automatically check for broken ones. Strip anything unnecessary (parameters, referrers, etc) off the end. As the base docs are in US English, any links should be to the English, desktop version of the site.
```

Fixed and fixed.

E.g. `example.com/page?tracking=fdsds&crap=fdssds&here=dsllds` -> `example.com/page`. Takes up less space, is less invasive of user privacy, looks cleaner and more trustworthy, and avoids any unintentional side effects.

- Use enumerated lists with ``#. `` to list items with a clearly defined order, otherwise, use bulleted lists with ``* ``. Include a space between the bullet/number and the text. No blank line between elements, unless a sublist is present, but blank lines around the list. Sublist should have no spaces around elements. You should try to avoid more than two levels of lists if at all possible, and keep lists to between 3 and 10-20 elements; anything shorter should be written in prose, and anything longer should be simplified or broken into multiple lists/sublists.
- Use ``note::`` for general callouts, ``warning::`` to warn users against procedures with potential serious consequences, and ``important::`` for key points that users don't want to miss. However, use these sparingly, only when clearly called for. Always make them separate blocks, with a blank line between the start of the block, like:
```
.. Note::

   This is a note.
```

## Images/Figures

> - Organize images into directories following the Google drive

Reviewer: What do you mean by this?

Author: ```suggestion
```

Its out of date now since the huge trache of screenshots I made is now obsolete with Spyder 4's release.

- Lower case file names and extensions, separate with hyphens, 3-character extensions

> - For ease of visibility, explanatory images should be full width; expository or those with very simple subjects can be half-width, right aligned

Reviewer: Why right aligned? Shouldn't they be left aligned?

Author: There may be another technical reason, but the layout flows better that way, and it makes conceptual sense for secondary, less important images to be such as they are supplementary and illustrative of the text, rather than the primary focus (or they'd be full width). That, and as I've emphasized multiple times its what is currently implemented in the docs; if you want to change it you can make an atomic PR after this one that changes it both here and in the docs themselves.

- Use PNG for all images, except photographs over 100-200 KB as a PNG (use moderately compressed JPEG); avoid sizes above 2000 x 2000 or below 1000 x 1000 if full width, or half of each if half width
- Always include alt text describing the content of the image
- No space before, one space ("|") after if full width

## Writing (All Languages)

- Follow the [OpenStack General Writing Guidelines](https://docs.openstack.org/doc-contrib-guide/writing-style/general-writing-guidelines.html)
- Maintain an explanatory rather than expository tone if appropriate for the language (i.e. in English, "How" rather than "What"; e.g. "Spyder can do XXX." -> "To do XXX, you can...")

> - Focus on the user; this is documentation, not promotional material

Reviewer: Good point!

Author: I sometimes forget this

- Try to use simple language if possible and avoid buzzwords, jargon and acronyms

> - Make sure capitalization, spacing etc. of proper names and acronyms are correct and consistent per the original (or the canonical localized form is availible)

Reviewer: * I don't understand what you mean by this.
* is available -> if available ?

Author: ```suggestion
- Make sure capitalization, spacing etc. of proper names and acronyms are correct and consistent per the original (or the canonical localized form if available)
```

Yup, fixed.

Your English language powers have doubled since we last met xD

- Must use ISO 8601 (YYYY-MM-DD HH:MM:SS, 24-hour format, with leading zeros) for all numeric dates and times. No convention is imposed for textual forms (e.g. January 1, 2018; 1 de abril de 2019; etc) so long as all three components are provided, the month names are written out in full, and the year is written as 4 digits.
-  ``.`` must be used as the decimal separator in languages that at least partially use it (i.e. a native speaker would not be unfamiliar with its use), and suggested otherwise at the option of the translator to be consistent with its use in Python and  internet standards, and avoid the textual ambiguity of `,`; regardless of which is chosen, it is more important to use it consistently throughout the docs of a given language
- If a separator is used between groups of digits for readability, it must be the space to avoid ambiguity per ISO standard conventions
- Should use only SI/metric for units

## Writing (English)

- Standard written English; US spellings are nominally the default for consistency but not required, and any given page should keep to only one convention. Avoid dialect-specific words.

> - AP style unless noted elsewhere here

Reviewer: What does this mean?

Author: Use AP style, unless stated otherwise by this style guide
